## EE 502 P: Analytical Methods for Electrical Engineering
    
# Homework 6: Fourier Analysis
## Due November 17, 2024 by 11:59 PM
### <span style="color: red">Enrique Antunano</span>

Copyright &copy; 2024, University of Washington

<hr>

**Instructions**: Please use this notebook as a template. Answer all questions using well formatted Markdown with embedded LaTeX equations, executable Jupyter cells, or both. Submit your homework solutions as an `.ipynb` file via Canvas.

<span style="color: red'">
Although you may discuss the homework with others, you must turn in your own, original work.
</span>

**Things to remember:**
- Use complete sentences. Equations should appear in text as grammatical elements.
- Comment your code.
- Label your axes. Title your plots. Use legends where appropriate.
- Before submitting a notebook, choose Kernel -> Restart and Run All to make sure your notebook runs when the cells are evaluated in order.


### 0. Warmup (Do not turn in)

- Make sure you download, read, and run the notebook for lecture 7. Work through the notebook cell by cell and see what happens when you change the expressions, and make up some of your own.
- The material covered in class is intended to be an introductory overview of the incredibly rich and expansive subject of Fourier Analysis.
- If you would like more resources, the following book is a good place to start. It is actually the notes used to teach intro signal processing at Stanford. Most of the material I covered in class is in these notes, plu a whole lot more.
> Osgood, [Lecture notes for EE 261](
https://see.stanford.edu/materials/lsoftaee261/book-fall-07.pdf).
- A more advanced and concise treatment of the subject is in the following book.
> Stein and Shakarchi, [Fourier Analysis: An Introduction](https://www.amazon.com/Fourier-Analysis-Introduction-Princeton-Lectures/dp/069111384X/ref=sr_1_1?keywords=stein+and+shakarchi+fourier&qid=1572974320&sr=8-1).

In [1]:
import math
import scipy # Has linear algebra
import scipy.ndimage
import numpy as np
import sympy as sm
sm.init_printing(use_latex='mathjax')
import matplotlib.pyplot as plt
import matplotlib.patches as pch
%matplotlib inline
from matplotlib.ticker import MaxNLocator # to make integer ticks on x axis

### 1. Exponential form of Complex Numbers

Express each of the following complex numbers in the form $r e^{i\theta}$. Do this by hand (or in Latex) and check your work with Python.

$$0$$
$$1$$
$$-1$$
$$i$$
$$1-2i$$
$$-8+7i$$

In [2]:
# Solve and plot complex vectors
r, theta = sm.symbols("r,theta")

complex_numbers     = [0,1,-1,sm.I,1-(2*sm.I),-8+(7*sm.I)]
abs_array           = []
arctan_array        = []

for x in range(len(complex_numbers)):
    abs_complex_num      = abs(complex_numbers[x])
    arctan_complex_num   = sm.arg(complex_numbers[x])

    abs_array.append(abs_complex_num)
    arctan_array.append(arctan_complex_num)

complex_numbers,abs_array, arctan_array

⎛                                                          ⎡           π       ↪
⎜[0, 1, -1, ⅈ, 1 - 2⋅ⅈ, -8 + 7⋅ⅈ], [0, 1, 1, 1, √5, √113], ⎢nan, 0, π, ─, -ata ↪
⎝                                                          ⎣           2       ↪

↪                    ⎤⎞
↪ n(2), π - atan(7/8)⎥⎟
↪                    ⎦⎠

1. 
a. Calculate the $\textbf{Magnitude}$ of $\textit{z}$, $$|z| = r = \sqrt{a^2 + b^2} = \sqrt{0^2 + 0^2} = 0$$
$$re^{\textit{i} \theta} = r(cos\theta+i sin \theta) = 0 * (cos\theta+i sin \theta) = 0$$

b. Calculate the $\textbf{Magnitude}$ of $\textit{z}$, $$|z| = r = \sqrt{a^2 + b^2} = \sqrt{1^2 + 0^2} = 1$$
Calculate $\textbf{argument}$ of $\textit{z}$, $$ \theta = arctan(y,x) = arctan(\frac{0}{1}) = 0$$
$$re^{\textit{i} \theta} = 1 * e^{\textit{i} * 0} = e^{0} = 1$$

c. Calculate the $\textbf{Magnitude}$ of $\textit{z}$, $$|z| = r = \sqrt{a^2 + b^2} = \sqrt{(-1)^2 + 0^2} = 1$$
Calculate $\textbf{argument}$ of $\textit{z}$, $$ \theta = arctan(y,x) = arctan(- \frac{0}{1}) = \pi$$
$$re^{\textit{i} \theta} = 1 * e^{\pi \textit{i} } = e^{\pi \textit{i} } = -1$$

d. Calculate the $\textbf{Magnitude}$ of $\textit{z}$, $$|z| = r = \sqrt{a^2 + b^2} = \sqrt{(0)^2 + 1^2} = 1$$
Calculate $\textbf{argument}$ of $\textit{z}$, $$ \theta = arctan(y,x) = arctan(1,0) \rightarrow \text{ Coordinates (0,1) for y>0: } \frac{\pi}{2}$$
$$re^{\textit{i} \theta} = 1 * e^{\frac{\pi}{2} \textit{i}} = e^{\frac{\pi}{2} \textit{i}} = \textit{i}$$

e. Calculate the $\textbf{Magnitude}$ of $\textit{z}$, $$|z| = r = \sqrt{a^2 + b^2} = \sqrt{(1)^2 + 2^2} = \sqrt{5}$$
Calculate $\textbf{argument}$ of $\textit{z}$, $$ \theta = arctan(y,x) = arctan(\frac{-2}{1})$$
Since angle is in the fourth quadrant, $$= - arctan(2)$$
$$ = - arctan(2) \approx -1.11 radians$$
$$re^{\textit{i} \theta} = \sqrt{5} e^{-1.11 \textit{i}}$$

f. Calculate the $\textbf{Magnitude}$ of $\textit{z}$, $$|z| = r = \sqrt{a^2 + b^2} = \sqrt{(-8)^2 + 7^2} = \sqrt{64+49} = \sqrt{113}$$
Calculate $\textbf{argument}$ of $\textit{z}$, $$ \theta = arctan(y,x) = arctan(\frac{7}{-8})$$
Since angle is in the second quadrant, $$= \pi - arctan(\frac{7}{8})$$
$$ arctan(\frac{7}{8}) \approx 0.7188 radians$$
$$= \pi - 0.7188 \approx 2.423$$
$$re^{\textit{i} \theta} = \sqrt{113} e^{2.423 \textit{i}}$$

### 2. A Trigonometric Identity

The goal of this exercise is to find formulas for $\cos(\alpha+\beta)$ and $\sin(\alpha + \beta)$. in terms of $\sin\alpha$, $\sin\beta$, $\cos\alpha$, and $\cos\beta$.

a) First, express $e^{(\alpha+\beta)i}$ directly in terms of the $\sin$ and $\cos$ using Euler's formula.

b) Next, express $e^{(\alpha+\beta)i}$ as a product of two exponentials. Then convert each term in the product into $\sin$ and $\cos$ using Euler's formula. Expand the expression and group the terms in the real part and the imaginary part.

c) Equate the real parts and imaginary parts from (a) and (b).

<br><br>
a) Convert from exponential form to trigonometric function form,
$$e^{(\alpha + \beta) \textit{i}} = cos(\alpha + \beta) + \textit{i } sin(\alpha + \beta)$$

b) First, express $e^{(\alpha + \beta) \textit{i}}$ as a product of exponentials,
$$ e^{(\alpha + \beta) \textit{i}} = e^{\alpha \textit{i}}e^{\beta \textit{i}}$$

Convert each term in the product from exponential form to trigonometric function form,
$$ e^{\alpha \textit{i}} = cos\alpha + \textit{i } sin\alpha$$
$$ e^{\beta \textit{i}} = cos\beta + \textit{i } sin\beta$$
$$ e^{\alpha \textit{i}}e^{\beta \textit{i}} = (cos\alpha + \textit{i } sin\alpha)(cos\beta + \textit{i } sin\beta)$$

Expand the expression,
$$ e^{\alpha \textit{i}}e^{\beta \textit{i}} = cos\alpha cos\beta + \textit{ i } cos\alpha sin\beta + \textit{ i } cos\beta sin\alpha - sin\alpha sin\beta$$

Group terms into real and imaginary parts,
$$ e^{\alpha \textit{i}}e^{\beta \textit{i}} = cos\alpha cos\beta - sin\alpha sin\beta + \textit{ i } (cos\alpha sin\beta + cos\beta sin\alpha)$$

c)
From $\textbf{part a}$, expand $e^{(\alpha + \beta) \textit{i}} = cos(\alpha + \beta) + \textit{i } sin(\alpha + \beta)$ in terms of $\sin\alpha$, $\sin\beta$, $\cos\alpha$, and $\cos\beta$ using the trigonometric angle sum and difference indentities,
$$ cos(\alpha + \beta) = \cos\alpha \cos\beta - \sin\alpha \sin\beta$$
$$ sin(\alpha + \beta) = \sin\alpha \cos\beta + \cos\alpha \sin\beta$$

Substitute these values back into $cos(\alpha + \beta) + \textit{i } sin(\alpha + \beta)$,
$$e^{(\alpha + \beta) \textit{i}} = \cos\alpha \cos\beta - \sin\alpha \sin\beta + \textit{i }(\sin\alpha \cos\beta + \cos\alpha \sin\beta)$$

The results from $\textbf{part a}$ and $\textbf{part b}$ are identical.
$$e^{(\alpha + \beta) \textit{i}} =  e^{\alpha \textit{i}}e^{\beta \textit{i}} = \cos\alpha \cos\beta - \sin\alpha \sin\beta + \textit{i }(\sin\alpha \cos\beta + \cos\alpha \sin\beta)$$

### 3. Fourier Series Approximation

Similar to the example of the square wave done in class, find a Fourier Series approximation to the function

$$
f(t) = \left |cos(2\pi t) \right|
$$

which is a *rectified cosine*. Make sure you get the period of this function correct. Make a plot with three subplots for $N=5$, $N=10$, and $N=20$ showing better and better approximations.

### 4. Orthogonality

Show that $\cos(2\pi t)$ and $\sin(2 \pi t)$ are orthogonal by showing that their inner product (a.k.a. dot product) is zero. Hint: Express the two functions in exponential form, and expand their product inside the integral.

### 5. Fourier Transforms

Find the Fourier Transforms of the following functions. You can use `sympy`.
Make sure to simplify, these shouldn't give complicated results.

Plot original functions and their transforms in a nice 3 $\times$ 2 grid of plots. Label the sub plots.

$$
f(t) = \left \{
\begin{array}{l}
1 \;\mathrm{if} \;t \geq 0 \\
0 \;\mathrm{else}
\end{array}
\right .
$$

$$
f(t) = \frac{1}{1+t^2}
$$

$$
f(t) = e^{-\pi t^2}
$$

### 6. Solve a system of ODEs

Similar to the ODE example in the notes, find a solution to the two differential equations

\begin{eqnarray}
\dot x & = & -y \\
\dot y & = & -x + u
\end{eqnarray}

a) First take the Fourier Transform of both equations.

b) Solve for $Y(s)$ in terms of $U(s)$.

c) Note that the solution for $Y(s)$ involves a function of $s$ times $U(s)$. Figure out the inverse Fourier Transform of that function.

d) Use a convolution to figure out $y(t)$ in terms of $u(t)$.

e) Find $X(s)$ in terms of $Y(s)$ and then $x(t)$ in terms of $y(t)$.

### 7.  Image Processing

Find a nice gray scale image to load, or use the Minion image from class. Call the resulting matrix $m$.

a) Render the image using `matplotlib`.

b) Take the DFT of the image $m$ to get a matrix $M$. Then make a new matrix $M_\mathrm{new}$ in which each entry is just the magnitude of the corresponding entry in $M$. Create a new image from the inverse Fourier transform of this image and render it.

c) Do the same thing, except set the entries in $M_\mathrm{new}$ to the angle of the entries in $M$. Create a new image from the inverse Fourier transform of this image and render it.

d) Do the same thing, except normalize all the entries in $M_\mathrm{new}$ to have magnitude 1. Create a new image from the inverse Fourier transform of this image and render it.

What might you conclude about the information contained in the magnitude of the transformed image? What might you conclude about the information contained in the angle?